### PR TITLE
Add support to multi-account environments

### PR DESCRIPTION
**Pull request by lakkeger**
# Motivation
tflocal is only usable currently with a single account as all resources are deployed into the default one. This behavior is not ideal to run tests or develop code for many production cases when terraform code references to resources outside of this account ie to manage a domain registration or validate certificates on a central DNS managing account.

Related issue: #29

# Changes
- Added new env variable to trigger look up for provided possible override values instead of the mocked access key
- Added tests regarding above functionality

# TODO
- [x] Trigger predefined access key usage for `CUSTOMISE_ACCESS_KEY=1`
- [x] Use AWS_ACCESS_KEY_ID environment variable
- [x] Use boto3 to get ACCESS_KEY_ID from default session
- [x] Use AWS_PROFILE or AWS_DEFAULT_PROFILE variable to get access key
- [x] Use access key from Terraform files
- [x] ~Import original arguments into the mock file~
- [x] Test above use cases and edge cases
- [x] Sync region and access key behavior

### CODE REVIEW

File: bin/tflocal [hunk 1]
```diff
@@ -243,6 +249,28 @@ def get_region() -> str:
...
+    access_key = str(os.environ.get("AWS_ACCESS_KEY_ID") or provider.get("access_key", "")).strip()
+    if access_key and access_key != DEFAULT_ACCESS_KEY:
+        # Change live access key to mocked one
+        return deactivate_access_key(access_key)
+    try:
+        # If boto3 is installed, try to get the access_key from local credentials.
+        # Note that boto3 is currently not included in the dependencies, to
+        # keep the library lightweight.
+        import boto3
+        profile = str(os.environ.get("AWS_PROFILE") or os.environ.get("AWS_DEFAULT_PROFILE", "default")).strip()
```
Comment (whummer) [inline bin/tflocal:262]: Should we leave the profile handling to botocore, instead of handling the cases directly here? Seems like both cases are already covered in the internal implementation here: https://github.com/boto/botocore/blob/9b401c75699033b9988c03bcd51f221751087c5c/botocore/configprovider.py#L54

I think we could get away with simply creating a boto3 session with default configs like below, or do you see an edge case where this would not achieve the desired behavior?
```
access_key = boto3.session.Session().get_credentials().access_key
```

File: tests/test_apply.py [hunk 2]
```diff
@@ -1,16 +1,179 @@
...
+import boto3
+
 THIS_PATH = os.path.abspath(os.path.dirname(__file__))
 ROOT_PATH = os.path.join(THIS_PATH, "..")
 TFLOCAL_BIN = os.path.join(ROOT_PATH, "bin", "tflocal")
 LOCALSTACK_ENDPOINT = "http://localhost:4566"
 
 
+def test_access_key_override_fallback(monkeypatch):
+    access_key = f"{mock_access_key()}"
```
Comment (whummer) [inline tests/test_apply.py:17]: nit: could be simplified to:
```suggestion
    access_key = mock_access_key()
```

File: tests/test_apply.py [hunk 3]
```diff
@@ -1,16 +1,179 @@
...
+    monkeypatch.setenv("AWS_ACCESS_KEY_ID", access_key)
+    bucket_name = f"{short_uid()}"
```
Comment (whummer) [inline tests/test_apply.py:19]:
```suggestion
    bucket_name = short_uid()
```

File: tests/test_apply.py [hunk 4]
```diff
@@ -1,16 +1,179 @@
...
+    assert bucket_name in s3_bucket_names
+
+    s3 = client("s3", region_name="eu-west-1", aws_access_key_id=access_key)
+    s3_buckets = s3.list_buckets().get("Buckets")
+    s3_bucket_names = [s["Name"] for s in s3_buckets]
+
+    assert bucket_name not in s3_bucket_names
+
+
+def test_access_key_override_by_env_var(monkeypatch):
```
Comment (whummer) [inline tests/test_apply.py:42]: nit: this test is quite similar to `test_access_key_override_fallback` above - perhaps we could consider combining into a single function, and parametrizing the test via `@pytest.mark.parametrize("customize_access_key", [True, False])`

(not critical at all though, just an idea for future iterations, no need to update the PR at this point)

File: bin/tflocal [hunk 5]
```diff
@@ -243,6 +249,28 @@ def get_region() -> str:
...
+        access_key = boto3.session.Session(profile_name=profile).get_credentials().access_key
+    except Exception:
+        pass
+    # fall back to default region
+    return deactivate_access_key(access_key or DEFAULT_ACCESS_KEY)
+
+
+def deactivate_access_key(access_key) -> str:
```
[the remark this answers is not in the document]
Comment (bentsku) [reply bin/tflocal:270]: Could also a add the type hints 👍 
```suggestion
def deactivate_access_key(access_key: str) -> str:
```

File: bin/tflocal [hunk 6]
```diff
@@ -118,6 +120,10 @@ def create_provider_config_file(provider_aliases=None):
     for provider in provider_aliases:
         endpoints = "\n".join([f'{s} = "{get_service_endpoint(s)}"' for s in services])
         provider_config = TF_PROVIDER_CONFIG.replace("<endpoints>", endpoints)
+        provider_config = provider_config.replace(
+            "<access_key>",
+            get_access_key(provider) if CUSTOMISE_ACCESS_KEY else DEFAULT_ACCESS_KEY
+        )
```
Comment (bentsku) [inline bin/tflocal:126]: this is a extremely small nit, but I would replace the `access_key` before replacing the `endpoints`, as it makes the string much bigger once the endpoints are in, and the search to replace would be longer. with a quick `timeit` checks, it seems it'd be around 30% slower.
```suggestion
        provider_config = TF_PROVIDER_CONFIG.replace(
            "<access_key>",
            get_access_key(provider) if CUSTOMISE_ACCESS_KEY else DEFAULT_ACCESS_KEY
        )
        endpoints = "\n".join([f'{s} = "{get_service_endpoint(s)}"' for s in services])
        provider_config = provider_config.replace("<endpoints>", endpoints)

```

We actually should be able to move the list comprehension to get the endpoints out of the loop and use it inside the loop, to not recalculate it every time when the result would be the same. 

Nothing blocking, this can be merged as is!

**Review by bentsku — APPROVED** [review]
LGTM! This is really great to see, and it's going to enable many more use cases. Thanks a lot for tackling this.

Same as Waldemar, especially around the British/American spelling of "customise", otherwise tiny nits, but Waldemar was already very thorough in his review 😄

File: tests/test_apply.py (hunk 4 above)
Comment (lakkeger) [reply tests/test_apply.py:42]: The cases are slightly different as the assertion is switching over so just introducing the parameterization is not enough here. In the next commit I'll add some sub-functions so we reduce the code repetition.

**Review by alexrashed — APPROVED** [review]
Awesome! This is a really cool new feature! And the code as well as the tests are looking great! 🚀 💯

**Comment by cshimmin** [comment]
As an interested observer from the US, I recommend either supporting the alternative spelling `CUSTOMIZE_ACCESS_KEY`, or avoid the ambiguity altogether with a different name. Perhaps `ACCESS_KEY_PASSTHROUGH`?

**Comment by alexrashed** [comment]
Hi @cshimmin!

> As an interested observer from the US, I recommend either supporting the alternative spelling CUSTOMIZE_ACCESS_KEY, or avoid the ambiguity altogether with a different name. Perhaps ACCESS_KEY_PASSTHROUGH?

Thanks for the comment! Actually, that was one of @whummer's PR review comments and @lakkeger already changed it in the code (but not yet in the PR description). 😛
